web3/_utils/contract_error_handling.py

```diff
@@ -0,0 +1,132 @@
+from eth_abi import (
+    abi,
+)
+from eth_utils import (
+    to_bytes,
+)
+
+from web3.exceptions import (
+    ContractCustomError,
+    ContractLogicError,
+    ContractPanicError,
+    OffchainLookup,
+)
+from web3.types import (
+    RPCResponse,
+)
+
+# func selector for "Error(string)"
+SOLIDITY_ERROR_FUNC_SELECTOR = "0x08c379a0"
+
+# --- CCIP Read - EIP-3668 --- #
+# the first 4 bytes of keccak hash (func selector) for:
+# "OffchainLookup(address,string[],bytes,bytes4,bytes)"
+OFFCHAIN_LOOKUP_FUNC_SELECTOR = "0x556f1830"
+OFFCHAIN_LOOKUP_FIELDS = {
+    "sender": "address",
+    "urls": "string[]",
+    "callData": "bytes",
+    "callbackFunction": "bytes4",
+    "extraData": "bytes",
+}
+
+
+# --- Solidity Panic Error, as of Solidity 0.8.0 --- #
+PANIC_ERROR_FUNC_SELECTOR = "0x4e487b71"
+PANIC_ERROR_CODES = {
+    "00": "Panic error 0x00: Generic compiler inserted panics.",
+    "01": "Panic error 0x01: Assert evaluates to false.",
+    "11": "Panic error 0x11: Arithmetic operation results in underflow or overflow.",
+    "12": "Panic error 0x12: Division by zero.",
+    "21": "Panic error 0x21: Cannot convert value into an enum type.",
+    "22": "Panic error 0x12: Storage byte array is incorrectly encoded.",
+    "31": "Panic error 0x31: Call to 'pop()' on an empty array.",
+    "32": "Panic error 0x32: Array index is out of bounds.",
+    "41": "Panic error 0x41: Allocation of too much memory or array too large.",
+    "51": "Panic error 0x51: Call to a zero-initialized variable of internal "
+    "function type.",
+}
+
+
+def raise_contract_logic_error_on_revert(response: RPCResponse) -> RPCResponse:
+    """
+    Reverts contain a `data` attribute with the following layout:
+        "Reverted "
+        Function selector for Error(string): 08c379a (4 bytes)
+        Data offset: 32 (32 bytes)
+        String length (32 bytes)
+        Reason string (padded, use string length from above to get meaningful part)
+
+    See also https://solidity.readthedocs.io/en/v0.6.3/control-structures.html#revert
+    """
+    if not isinstance(response["error"], dict):
+        raise ValueError("Error expected to be a dict")
+
+    data = response["error"].get("data", "")
+
+    # Ganache case:
+    if isinstance(data, dict) and response["error"].get("message"):
+        raise ContractLogicError(
+            f'execution reverted: {response["error"]["message"]}', data=data
+        )
+
+    # Parity/OpenEthereum case:
+    if data.startswith("Reverted "):
+        # "Reverted", function selector and offset are always the same for revert errors
+        prefix = f"Reverted {SOLIDITY_ERROR_FUNC_SELECTOR}"
+        data_offset = f"{'00' * 31}" + "20"  # 0x0000...0020 (32 bytes)
```

Comment:
nit: you don't need the `f-string` here
```suggestion
        data_offset = "00" * 31 + "20"  # 0x0000...0020 (32 bytes)
```